### PR TITLE
Keep omega within [-1, 1] bounds

### LINKED ISSUE #5064: `omega` gives values that are not within [-1,1] bounds
According to the paper this metric has been based upon, [found here](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3604768/) , omega needs to be between [-1,1].

However nx.omega gives as a result many values that are not within [-1,1] bounds

### Current Behavior

values not being within the expected [-1,1] bounds

### Expected Behavior

values should be within [-1,1] bounds

### Steps to Reproduce

```python
import networkx as nx
G = nx.barbell_graph(5,1)
print(nx.omega(G,seed=1)) # result is -1.505815972222222
```
### Environment

Python version: 3.7.4
NetworkX version: 2.5.1

### PR DESCRIPTION
**Pull request by fabriz-io**
Closes #5064 

The problem of #5064 regards network examples, where the calculated small-world omega metric does not lie within the bounds of the specified [-1,1] intervall. In order to check where this is coming from, I tried to find any differences in the description of the calculation from the original paper to the implemented version. The following was changed in this PR:

- Changing from nx.transitivity to nx.clustering for the clustering coefficient
- Changing default values for niter and nrand. Note: niter=5 is specified for the lattice and niter=10 is specified for the random reference. And nrand_random=50 is mentioned in the section "Random and lattice network construction" for the random reference graph. nrand_lattice is not explicitly stated, which is why I leave the default value for nrand=10, as the computation of omega quickly becomes unfeasible, mostly due to the lattice calculation.
- Changing from average clustering to maximal clustering: In the original paper they only accept a new lattice reference graph, if the clustering in the generated graph is higher than that from the previos generated lattice. The aim of this procedure is to find a lattice reference with the maximum clustering while preserving the degree distribution of the original graph.
- Adding tests for some graphs to check if there lie within the bounds. I used the graph from the issue, the already used graphs in the testing function and added the karate graph. In a previous version I had various additional graphs, but the calculation took quite a while, which is why I kept the graph examples to a minimum in order to not blew up the test suite execution time.

### CODE REVIEW

File: networkx/algorithms/smallworld.py [hunk 1]
```diff
@@ -354,15 +353,31 @@ def omega(G, niter=100, nrand=10, seed=None):
...
-        Gr = random_reference(G, niter=niter, seed=seed)
-        Gl = lattice_reference(G, niter=niter, seed=seed)
-        randMetrics["C"].append(nx.transitivity(Gl))
+
+    # Calculate initial average clustering coefficient which potentially will
+    # get replaced by higher clustering coefficients from generated lattice
+    # reference graphs
+    Cl = nx.average_clustering(G)
+
+    niter_lattice_reference = niter
```
Comment (fabriz-io) [inline networkx/algorithms/smallworld.py:362]: The niter_lattice_reference=5 default is taken from the upon based paper. As this takes significantly longer to calculate as the random reference, I increased the niter_random_reference to be closer to the previous default. In my opinion there is no natural to choose an appropriate value, I just chose the ones which where computationally feasible for some testes networks. Setting both niters to the same value might be fine too.

File: networkx/algorithms/smallworld.py [hunk 2]
```diff
@@ -354,15 +353,31 @@ def omega(G, niter=100, nrand=10, seed=None):
...
+    niter_random_reference = niter * 2
```
Comment (fabriz-io) [inline networkx/algorithms/smallworld.py:363]: The niter_lattice_reference=5 default is taken from the upon based paper. As this takes significantly longer to calculate as the random reference, I increased the niter_random_reference to be closer to the previous default. In my opinion there is no natural to choose an appropriate value, I just chose the ones which where computationally feasible for some testes networks. Setting both niters to the same value might be fine too.

File: networkx/algorithms/tests/test_smallworld.py [hunk 3]
```diff
@@ -58,3 +58,15 @@ def test_omega():
...
     assert omegal < omegas and omegas < omegar
+
+    # Test that omega lies within the [-1, 1] bounds
+    G_barbell = nx.barbell_graph(5, 1)
+    G_karate = nx.karate_club_graph()
+
+    omega_barbell = nx.omega(G_barbell)
+    omega_karate = nx.omega(G_karate, nrand=2)
+
+    omegas = (omegal, omegar, omegas, omega_barbell, omega_karate)
```
Comment (fabriz-io) [inline networkx/algorithms/tests/test_smallworld.py:69]: More graphs could be tested, but this takes already some time.

File: networkx/algorithms/smallworld.py [hunk 4]
```diff
@@ -181,12 +181,12 @@ def lattice_reference(G, niter=1, D=None, connectivity=True, seed=None):
             D[v, :] = D[nnodes - v - 1, :][::-1]
 
     niter = niter * nedges
-    ntries = int(nnodes * nedges / (nnodes * (nnodes - 1) / 2))
-    swapcount = 0
+    # maximal number of rewiring attempts per 'niter'
+    max_attempts = int(nnodes * nedges / (nnodes * (nnodes - 1) / 2))
```
Comment (fabriz-io) [inline networkx/algorithms/smallworld.py:185]: I suggest this as a more meaningful name for the variable

File: networkx/algorithms/smallworld.py [hunk 5]
```diff
@@ -181,12 +181,12 @@ def lattice_reference(G, niter=1, D=None, connectivity=True, seed=None):
...
 
-    for i in range(niter):
+    for _ in range(niter):
```
Comment (fabriz-io) [inline networkx/algorithms/smallworld.py:187]: varible i is unused

File: networkx/algorithms/smallworld.py [hunk 6]
```diff
@@ -181,12 +181,12 @@ def lattice_reference(G, niter=1, D=None, connectivity=True, seed=None):
             D[v, :] = D[nnodes - v - 1, :][::-1]
 
     niter = niter * nedges
-    ntries = int(nnodes * nedges / (nnodes * (nnodes - 1) / 2))
-    swapcount = 0
```
Comment (fabriz-io) [inline networkx/algorithms/smallworld.py:185]: Variable swapcount is unused

**Comment by dschult** [comment]
Can you summarize what you have changed and how that fixes the issue?
It looks like the substantive change is to compute the maximal clustering instead of the average clustering.
But it would help a lot to provide more description of what you did and why.
Thanks!!

**Comment by fabriz-io** [comment]
Sure! I included a description in the first comment of this PR, hope this is the right way @dschult  :)

**Comment by dschult** [comment]
Thanks for the summary above. That is helpful.

I have a couple of questions:
- Why did you replace the transitivity of G with the average clustering of G?
-  Why did you replace the `Cl` value with the maximum of average clustering of the `nrand` lattices instead of the average of them?

If these are from something n the article can you say where in the article it says to do that?
Thanks!

**Comment by dschult** [comment]
OK... I think everything I ask about in my previous post is answered [in your comment](https://github.com/networkx/networkx/issues/5064#issuecomment-975284052) from #5064 
I have no more questions at the moment -- I will play with it a little before officially approving it.
Other reviewers should consider this OK'd from my perspective. I don't expect any major issues.

**Comment by dschult** [comment]
It appears that omega is **not** bounded between [-1, 1].
Both the old version and the new version are consistently above 1 for the graph shown below. Similar code produces graphs that give omega less than -1.  It may be that the number of nodes is too small and in some limit the results get close to 0. But if that is true, then it shows that the constraint on omega is not a hard one.

I'm not sure which of these versions is more correct. But since the definition is fairly vague, maybe it doesn't matter too much which definition we use. I think we should change the doc_string to indicate that omega returns a real value, but not in [-1, 1] as sometimes suggested. Thoughts?

 Here's the test code:
```python
In [64]: all_o = []
    ...: N=14
    ...: p=.2
    ...: G=nx.binomial_graph(N, p, seed=22)
    ...: assert nx.is_connected(G)
    ...: for s in range(21,30):
    ...:   try:
    ...:     omega=nx.omega(G, seed=s)
    ...:     print(f"{s}: {omega}")
    ...:     all_o.append(omega)
    ...:   except:
    ...:     pass
    ...: print(max(*all_o), min(*all_o))
```
The old code gives:
```
21: 1.1536697247706422
22: 1.1302752293577982
23: 1.1302752293577982
24: 1.1211009174311928
25: 1.1385321100917432
26: 1.0871559633027525
27: 1.1564220183486238
28: 1.1183486238532112
29: 1.1481651376146789
1.1564220183486238 1.0871559633027525
```
With the new code:
```
21: 1.1788990825688075
22: 1.1082568807339452
23: 1.1440366972477065
24: 1.1412844036697247
25: 1.1385321100917432
26: 1.1224770642201836
27: 1.1674311926605505
28: 1.1481651376146789
29: 1.1403669724770642
1.1788990825688075 1.1082568807339452
```

**Comment by fabriz-io** [comment]
Sorry for the late reply, did not manage to look into this earlier. Thanks for pointing to this cases, thought I tested it enough and did not find any values outside the bounds anymore.

My feeling is similar, I think the out-of-the-bounds cases are mainly due to small graphs, where the path lengths just can't be that long. As a suggestion: What if we just contact the authors to get a definite answer? I could take over this at some point in the next days.

**Comment by dschult** [comment]
That sounds good to me.  Thanks!!

**Review by dschult — APPROVED** [review]
I updated the doc_string and did a few trials to compare values and timing. And I added a note to release_dev.

Thanks @fabriz-io !!

File: doc/release/release_dev.rst [hunk 7]
```diff
@@ -36,6 +36,8 @@ X contributors. Highlights include:
 Improvements
 ------------
 
+- [`#5216 <https://github.com/networkx/networkx/pull/5216>`_]
+  Make ``omega()`` closer to the published algorithm. The value changes slightly.
```
Comment (rossbar) [inline doc/release/release_dev.rst:40]: I think we should also mention that the default value for the `niter` parameter is changing from 1->5 in the case of `lattice_reference` and from 100->5 in the case of `omega`

**Review by rossbar — APPROVED** [review]
LGTM - just one comment about adding a little more info to the release note!
